IterableMatchers factory methods and ValueObject test support class. …

lemonboston opened the pull request:
…Move TestDoubles and FailAnswer to packages. #23 

---

@dmfs I am curios what you think about the added `ValueObject`. `String` may work instead but the pool always makes it less clear for me whether there is object or reference equality in place. And for `Matcher`s the difference can be important. So that was the motivation, to make it explicit.

dmfs commented:
Is there anything in this PR which we agreed on using? We really should have something like "named mocks" instead.

lemonboston commented:
One question is that you don't see the matching for `Iterable<T>` and not for `T...` a valid use-case? So you think we won't need that? (That's missing from hamcrest)

Regardless of that, I think `ValueObject` can still be useful for `Matcher` tests, like here. So it's a different thing from those name mocks, those are aimed at testing production code.
`ValueObject` can be used when we want to make sure the `Matcher` under test is comparing object equality and not reference equality. In other words, that the default `Matcher` it uses underneath for the elements is `isEqualTo()` and not `isSameInstance()`.

lemonboston commented on test-utils/src/main/java/org/dmfs/testutils/matchers/IterableMatchers.java:
```diff
@@ -0,0 +1,77 @@
+     * {@link Matcher} that matches when the actual {@link Iterable} has the same - meaning equals() is true - elements in same order as the provided expected items.
+     */
+    @Factory
+    public static <E> Matcher<Iterable<? extends E>> hasElements(E... items)
```
Yes, it's the same with a different method name. 'contains' implies the 'Collection.contains()' meaning for me, so even though I got used to it by now that hamcrest uses it differently, I figured if we had our `IterableMatcher` anyway why not find a better name, and use that consistently for all the overloaded factories.

dmfs replied:
I see. I didn't use `Collection.contains()` for a long time and got used to the name. FYI, there is an issue for this: https://github.com/hamcrest/JavaHamcrest/issues/140
IMO, the name `hasElements` still doesn't really imply that is has **only** the given elements. How about `iterates` or `iteratesTo`? Especially the latter one implies completeness to me.

lemonboston commented:
I've removed, reverted everything except the `IterableMatcher` factory methods.
Could you please check if it makes sense? I'd write the unit test then.
`iteratesTo()` sounds good to me, too, so I used that.

The problem is that we cannot have both `Iterable<Matcher<E>>` and `Iterable<E>` because of the same erasure. Not sure which one to keep. With `Matcher` we usually enumerate individually, but that means we use the vararg version anyway. So would `Iterable<e>` be used more often?
Or add both with slightly different name?

lemonboston commented:
How about? 
```java
iterateAs(Iterable<Matcher<E>> matchers)
iterateTo(Iterable<E> items)
```
